form/src/Components/Form.js

```diff
@@ -0,0 +1,120 @@
+import "../App.css";
+import React, { useState } from "react";
+import {
+  intro_words,
+  letter_content,
+  farewell,
+} from "../DefaultData/ChristmasPageData";
+
+const Form = ({ storage, firestore, user }) => {
+  const [files, setFiles] = useState([]);
+  const [urls, setUrls] = useState([]);
+  const [openingText, setOpeningText] = useState(intro_words.msg);
+  const [from, setFrom] = useState(user.displayName);
+  const [to, setTo] = useState(intro_words.to);
+  const [letterGreeting, setLetterGreeting] = useState(letter_content.greeting);
+  const [letterText, setLetterText] = useState(letter_content.msg);
+  const [closingText, setClosingText] = useState(farewell.msg);
+  const [signOff, setSignOff] = useState(farewell.from);
+
+  const handleSubmit = async (event) => {
+    event.preventDefault();
+    const urlTemp = await handleUpload();
+    setUrls([...urlTemp]);
+    handleFireStoreWrite();
+  };
+
+  const handleFireStoreWrite = () => {
+    console.log("Writing to firestore");
+    firestore
+      .collection("SiteData")
+      .add({
+        createdBy: user.email,
+        openingText,
+        from,
+        to,
+        letterGreeting,
+        letterText,
+        closingText,
+        signOff,
+        images: urls,
+      })
+      .then((docRef) => {
+        console.log("Document written with ID: ", docRef.id);
+      })
+      .catch((error) => {
+        console.error("Error adding document: ", error);
+      });
+  };
+
+  const handleUpload = async () => {
+    console.log("Uploading Files to storage bucket...");
+    const urltemp = await [...files].map(async (file) => {
+      const snapshot = await storage
+        .ref(user.displayName)
+        .child(`${file.name}`)
+        .put(file);
+      console.log(
+        `Uploade Successful!!! 
+        \nBytes Transferred:${snapshot.bytesTransferred}`
+      );
+      const url = await storage
+        .ref(user.displayName)
+        .child(file.name)
+        .getDownloadURL();
+      return url;
+    });
+    return Promise.all(urltemp).then((values) => values);
```

Comment:
what does the `then((values) => values);` part do?

Reply:
the urltemp is an array of promises that need to be resolved before leaving the function. Otherwise, the form will submit with few to no URLs. Promise.all basically waits till all of them are being resolved first.